File: VoiSlateParser/Models/SlateLogItem.cs
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using VideoTimecode;

using VoiSlateParser.Helper;
// ReSharper disable InconsistentNaming

namespace VoiSlateParser.Models;

public enum TkStatus
{
    notChecked = 0,
    ok = 1,
    bad = 2
}

public enum ShtStatus
{
    notChecked = 0,
    ok = 1,
    nice = 2
}

public class SlateLogItem
{
    public SlateLogItem(
        string scn,
        string sht,
        int tk,
        string filenamePrefix,
        string filenameLinker,
        int filenameNum,
        string tkNote,
        string shtNote,
        string scnNote,
        List<string> trackList,
        TkStatus okTk = TkStatus.notChecked,
        ShtStatus okSht = ShtStatus.notChecked
        )
    {
        this.scn = scn;
        this.sht = sht;
        this.tk = tk;
        this.filenamePrefix = filenamePrefix;
        this.filenameLinker = filenameLinker;
        this.filenameNum = filenameNum;
        this.tkNote = tkNote;
        this.shtNote = shtNote;
        this.scnNote = scnNote;
        this.trackList = trackList;
        this.okTk = okTk;
        this.okSht = okSht;
        this.bwfSynced = false;
        this.startTc = new Timecode(0,FrameRate.FrameRate24);
        this.endTc = new Timecode(0,FrameRate.FrameRate24);
        this.fileLength = new Timecode(0,FrameRate.FrameRate24);
        this.ubits = "00000000";
        this.bwfList = new();
    }

    public string fileName => filenamePrefix + filenameLinker + filenameNum.ToString().PadLeft(3,'0');
    public string scn { get; set; }
    public string sht { get; set; }
    public int tk { get; set; }
    public string filenamePrefix { get; set; }
    public string filenameLinker { get; set; }
    public int filenameNum { get; set; }
    public string tkNote { get; set; }
    public string shtNote { get; set; }
    public string scnNote { get; set; }
    public List<string>? trackList{ get; set;}

    // public string scnNote 
    // {
    //     get => shtNote;
    //     set => scnNote = EncodingHelper.ConvertUtfToIso(value);
    // }

    public TkStatus okTk { get; set; }
    public ShtStatus okSht { get; set; }
    
    public Timecode startTc { get; set; }
    public string startTcString => startTc.ToString();
    public Timecode endTc { get; set; }
    public string endTcString => endTc.ToString();
    public Timecode fileLength { get; set; }
    public string fileLengthString => fileLength.ToString();
    
    public List<FileInfo> bwfList { get; set; }
    public EnumerableRowCollection<DataRow> videoList { get; set; }
    public List<string?> videoListNames => (videoList != null)? videoList.Select(row => row["File Name"].ToString()).ToList():new();
    public string ubits { get; set; }

    public bool bwfSynced { get; set; }
    public bool videoSynced { get; set; }

    public bool Contains(string filterText)
    {
        if (scn.Contains(filterText)) return true;
        if (sht.Contains(filterText)) return true;
        if (tkNote.Contains(filterText)) return true;
        if (shtNote.Contains(filterText)) return true;
        if (scnNote.Contains(filterText)) return true;
        if (fileName.Contains(filterText)) return true;
        if (bwfList != null)
        {
            if (BwfContainsFileter(filterText)) return true;
        }
        return false;
    }

    private bool BwfContainsFileter(string filterText)
    {
    // a function to find if any items in bwfList.Name contains fileterText
    // if so, return true, else false.
    // if bwfList is null, return true.
    // if bwfList is empty, return true.
    // if bwfList is not empty, return true if any item contains filterText.
        if (bwfList == null)
        {
            {
                return true;
            }
        }

        if (bwfList != null)
        {
            foreach (var bwf in bwfList)
            {
                if (bwf.Name.Contains(filterText))
                {
                    {
                        return true;
                    }
                }
            }
        }

        return false;
    }


}

File: src/lib/models/slate_log_item.dart
import 'package:hive_flutter/hive_flutter.dart';
import 'package:dart_json_mapper/dart_json_mapper.dart';

part 'slate_log_item.g.dart';

/// The status of the file.
/// * 0: not checked
/// * 1: ok
/// * 2: bad
@JsonSerializable()
@HiveType(typeId: 4)
enum TkStatus {
  @HiveField(0)
  notChecked,
  @HiveField(1)
  ok,
  @HiveField(2)
  bad,
}

/// The status of the slate.
/// * 0: not checked
/// * 1: ok
/// * 2: nice
@JsonSerializable()
@HiveType(typeId: 5)
enum ShtStatus {
  @HiveField(0)
  notChecked,
  @HiveField(1)
  ok,
  @HiveField(2)
  nice,
}

@JsonSerializable()
@HiveType(typeId: 6)
class SlateLogItem {
  @HiveField(0)
  String scn;
  @HiveField(1)
  String sht;
  @HiveField(2)
  int tk;
  @HiveField(3)
  String filenamePrefix;
  @HiveField(4)
  String filenameLinker;
  @HiveField(5)
  int filenameNum;
  String get fileName {
    return filenamePrefix +
        filenameLinker +
        filenameNum.toString().padLeft(3, '0');
  }

  @HiveField(6)
  String tkNote;
  @HiveField(7)
  String shtNote;
  @HiveField(8)
  String scnNote;
  @HiveField(9)
  TkStatus okTk;
  @HiveField(10)
  ShtStatus okSht;

  SlateLogItem({
    required this.scn,
    required this.sht,
    required this.tk,
    required this.filenamePrefix,
    required this.filenameLinker,
    required this.filenameNum,
    required this.tkNote,
    required this.shtNote,
    required this.scnNote,
    @JsonProperty(name: 'okTk') TkStatus currentOkTk = TkStatus.notChecked,
    @JsonProperty(name: 'okSht') ShtStatus currentOkSht = ShtStatus.notChecked,
  })  : okTk = currentOkTk,
        okSht = currentOkSht;
}
